timely_beliefs/beliefs/utils.py

```diff
@@ -1030,3 +1038,68 @@ def meta_repr(
             for attr in tb_structure._metadata
         ]
     )
+
+
+def upsample_beliefs_data_frame(
+    df: "classes.BeliefsDataFrame" | pd.DataFrame,
+    event_resolution: timedelta,
+    keep_nan_values: bool = False,
+) -> "classes.BeliefsDataFrame":
+    """Because simply doing df.resample().ffill() does not correctly resample the last event in the data frame.
+
+    :param df:                  In case of a regular pd.DataFrame, make sure to set df.event_resolution before passing it to this function.
+    :param event_resolution:    Resolution to upsample to.
+    :param keep_nan_values:     If True, place back resampled NaN values. Drops NaN values by default.
+    """
+    from_event_resolution = df.event_resolution
+    resample_ratio = pd.to_timedelta(to_offset(from_event_resolution)) / pd.Timedelta(
+        event_resolution
+    )
+    if keep_nan_values:
+        # back up NaN values
+        unique_event_value_not_in_df = df["event_value"].abs().sum() + 1
+        df = df.fillna(unique_event_value_not_in_df)
+    new_index = initialize_index(
+        start=df.index[0],
```

Comment:
I think this would break in case of having an empty DataFrame. In that case, I would just change the resolution and return the same object perhaps, complaining with a warning.